cephfs/directory.go

```diff
@@ -0,0 +1,131 @@
+package cephfs
+
+/*
+#cgo LDFLAGS: -lcephfs
+#cgo CPPFLAGS: -D_FILE_OFFSET_BITS=64
+#include <stdlib.h>
+#include <dirent.h>
+#include <cephfs/libcephfs.h>
+*/
+import "C"
+
+import (
+	"unsafe"
+)
+
+// Directory represents an open directory handle.
+type Directory struct {
+	mount *MountInfo
+	dir   *C.struct_ceph_dir_result
+}
+
+// OpenDir returns a new Directory handle open for I/O.
+//
+// Implements:
+//  int ceph_opendir(struct ceph_mount_info *cmount, const char *name, struct ceph_dir_result **dirpp);
+func (mount *MountInfo) OpenDir(path string) (*Directory, error) {
+	var dir *C.struct_ceph_dir_result
+
+	cPath := C.CString(path)
+	defer C.free(unsafe.Pointer(cPath))
+
+	ret := C.ceph_opendir(mount.mount, cPath, &dir)
+	if ret != 0 {
+		return nil, getError(ret)
+	}
+
+	return &Directory{
+		mount: mount,
+		dir:   dir,
+	}, nil
+}
+
+// Close the open directory handle.
+//
+// Implements:
+//  int ceph_closedir(struct ceph_mount_info *cmount, struct ceph_dir_result *dirp);
+func (dir *Directory) Close() error {
+	return getError(C.ceph_closedir(dir.mount.mount, dir.dir))
+}
+
+// Inode represents an innode number in the file system.
+type Inode uint64
+
+// DirEntry represents an entry within a directory.
+type DirEntry struct {
+	Inode Inode
+	Name  string
+}
+
+// DirEntries provides a convenient wrapper around slices of DirEntry items.
+// For example, use the Names() call to easily get only the names from a
+// DirEntry slice.
+type DirEntries []*DirEntry
+
+// toDirEntry converts a c struct dirent to our go wrapper.
+func toDirEntry(de *C.struct_dirent) *DirEntry {
+	return &DirEntry{
+		Inode: Inode(de.d_ino),
+		Name:  C.GoString(&de.d_name[0]),
+	}
+}
+
+// ReadDir reads a single directory entry from the open Directory.
+// A nil DirEntry pointer will be returned when the Directory stream has been
+// exhausted.
+//
+// Implements:
+//  int ceph_readdir_r(struct ceph_mount_info *cmount, struct ceph_dir_result *dirp, struct dirent *de);
+func (dir *Directory) ReadDir() (*DirEntry, error) {
```

Comment:
I think it would be better to stick to the API as defined by the "os" package. If CephFS implements all File functions, it can be used like a normal local filesystem (except for a few operations like Open, I guess).

Reply:
Could you be more specific about what calls in that package you wan to mimic? (pun intended)
I think I am going to object to that idea but I want to make sure we're discussing the right thing first.

Reply:
I was thinking of https://golang.org/pkg/os/#File.Readdir (note the lowecase "dir" and arguments) and instead of DirEntry returning a FileInfo interface.

Readdir() would map to ceph_readdirplus_r() with some loop, and Readdirnames() maps to ceph_readdir_r().

This does divert from the "keep the API close to libcephfs", but on the other hand, there is good standardization for filesystem APIs. I think it will be easier for consumers of go-ceph if the API for CephFS is close to the usual filesystem API.

Reply:
As you may have suspected I would much prefer to expose the ceph apis more directly. Plus, I really dislike the FileInfo type [1]. However, as you note we can add a compatibility layer with the Go filesystem stuff if we want later. I can see tweaking the the convenience functions I've added in the final patch to make it easier to make this compatibility though. For example ListN(n int) instead of the current List(). 
But I am going to object strongly to doing file info at this time. What I'm open to is naming the  fields differently than they are currently so we could add methods to the DirEntry type such that one could define interfaces that work for both FileInfo and DirEntry.

1 - To fill in that structure on a file system that doesn't have a call like readdirplus Go stats every direntry coming back from the readdir. Yuck!

Reply:
Ping: I'd like to continue this discussion when you have the time.

Reply:
yeah, I am ok with following the libcephfs functions and not a full Golang filesystem API. It is also difficult to do, as not all required interfaces are available from the os package.